dali/pipeline/data/dynamic_tensor_view.h

```diff
@@ -0,0 +1,623 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_DATA_DYNAMIC_TENSOR_VIEW_H_
+#define DALI_PIPELINE_DATA_DYNAMIC_TENSOR_VIEW_H_
+
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include "dali/core/tensor_shape.h"
+#include "dali/core/tensor_view.h"
+#include "dali/pipeline/data/types.h"
+
+namespace dali {
+
+/**
+ * @brief Tag type indicating that the TensorView is keeping the type information in runtime.
+ *
+ * It can be converted to the statically-typed TensorView using `view<T, ndim>()` function
+ * or by `TensorView::to_static_type<T, ndim>()` member function.
+ */
+struct DynamicType {
+  DynamicType() = delete;
+};
+
+template <typename Backend, typename DataType, int ndim = DynamicDimensions>
+struct DynamicTensorViewBase {
+  static_assert(std::is_same<std::remove_const_t<DataType>, void>::value,
+                "The underlying type must either be void or const void");
+  using element_type = DataType;
+  int dim() const {
+    return shape.sample_dim();
+  }
+  DALIDataType type() const {
+    return type_id;
+  }
+
+  ptrdiff_t num_elements() const {
+    return shape.num_elements();
+  }
+
+  /**
+   * @brief Utility to calculate pointer to element at given coordinates
+   */
+  template <typename... Indices>
+  DataType *operator()(int64_t idx0, Indices &&...idx) const {
+    return data + CalcOffset(shape, std::array<ptrdiff_t, sizeof...(Indices) + 1>{
+                                        idx0, (ptrdiff_t{idx})...});
+  }
+
+  /**
+   * @brief Utility to calculate pointer to element at given coordinates
+   */
+  template <typename Offset>
+  DataType *operator()(const Offset &pos) const {
+    return data + CalcOffset(shape, pos);
+  }
+
+  DataType *data = nullptr;
+  TensorShape<ndim> shape = {};
+  DALIDataType type_id = DALI_NO_TYPE;
+
+ protected:
+  DynamicTensorViewBase() = default;
+  DynamicTensorViewBase(const DynamicTensorViewBase &) = default;
+  DynamicTensorViewBase(DataType *data, const TensorShape<ndim> &shape, DALIDataType type_id)
+      : data(data), shape(shape), type_id(type_id) {}
+  DynamicTensorViewBase(DataType *data, TensorShape<ndim> &&shape, DALIDataType type_id)
+      : data(data), shape(std::move(shape)), type_id(type_id) {}
+};
+
+template <typename Backend, int ndim = DynamicDimensions>
+struct DynamicTensorView : DynamicTensorViewBase<Backend, void, ndim> {
+  using Base = DynamicTensorViewBase<Backend, void, ndim>;
+
+  DynamicTensorView() = default;
+
+  /**
+   * @name Construct the view inferring the type_id from the pointer value.
+   */
+  // @{
+  template <typename T, typename = std::enable_if_t<!std::is_const<T>::value>>
+  DynamicTensorView(T *data, const TensorShape<ndim> &shape)
+      : Base(data, shape, TypeTable::GetTypeId<T>()) {}
```

Comment:
How about:
```suggestion
  template <typename T>
  DynamicTensorView(T *data, const TensorShape<ndim> &shape)
      : DynamicTensorView(static_cast<void*>(data,) shape, TypeTable::GetTypeId<T>()) {}
```
?
This will scream that a pointer of type `const something` cannot be converted to `void`. The error produced by SFINAE would be much harder to understand.